### PR TITLE
Cleanroom re-implementations for deleted code

### PR DESCRIPTION
**Pull request by kdkasad**
I attempted to re-implement all of the code (outside of `ui.go`) that was marked as needing re-implementation.

I only used the *TODO* comments and what I can discover just from using NoiseTorch to figure out what the missing code blocks are supposed to do.

### CODE REVIEW

**Comment by kdkasad** [comment]
While re-implementing the deleted UI logic, I noticed that long error messages get clipped instead of wrapping. I tried to fix this by using `w.LabelWrap(...)` instead of `w.Label(...)` but then I discovered a bug in nucular's implementation of `LabelWrap`.

So until aarzilli/nucular#69 is fixed, the UI will have to clip long error messages instead of wrapping them.

**Comment by kdkasad** [comment]
> Any reason not to use the predefined vars for colors in LabelColored calls?

I didn't notice those. I'll change that soon.

**Comment by kdkasad** [comment]
> While re-implementing the deleted UI logic, I noticed that long error messages get clipped instead of wrapping. I tried to fix this by using `w.LabelWrap(...)` instead of `w.Label(...)` but then I discovered a bug in nucular's implementation of `LabelWrap`.
> 
> So until [aarzilli/nucular#69](https://github.com/aarzilli/nucular/issues/69) is fixed, the UI will have to clip long error messages instead of wrapping them.

Turns out this was wrong and I was just using the function wrong. (See aarzilli/nucular#69 for details.)

Commit 2a7ec81 implements wrapped text. This isn't technically a required part of the code re-writing and could probably stand to be its own PR, but since it's related to other changes in this PR, I'm including it here.

**Comment by TheDukeofErl** [comment]
Not sure what's going on with the requesting of a review from someone resulting in a removal from a request as well but it looks like everyone is still in the requests list...

**Review by AXDOOMER — CHANGES_REQUESTED** [review]
Hi. 

Thank you for contributing to the reimplementation of problematic code in NoiseTorch. You rewrote all the code that I labeled as "must be rewritten from scratch" so far, so if you'd like to continue, I'll proceed with removing some of the Pipewire code from NoiseTorch's source code that must be rewritten. It's important that you do not look at the Pipewire code and that you have not already looked at it so that it can be a clean room reimplementation. 

I've left a few comments. Two of them contain one or more questions. One requires a change.

File: cli.go
Comment (kdkasad) [reply cli.go:32]: Most programs use a `-v` or `--verbose` flag to enable info-level logging. So it's more in line with many other command-line utilities. I was initially confused by the `-log` option as I thought it would take a filename argument and write logs to that file.

I didn't think about the fact that changing this would break existing scripts when I did it. I'll change the option back to `-log` instead of `-v`.

File: cli.go
Comment (kdkasad) [reply cli.go:188]: Is that a problem? Since all `removeLib` does is delete the file, I could change that to just call `os.Remove` instead.

File: module.go
Comment (kdkasad) [reply module.go:192]: I vote to leave this the way it is, since I don't actually know Go and I can't figure out how to get the error code out of the `err` variable. :)

And really we should display any error that occurs anyways, as there could maybe be others?

File: ui.go
Comment (kdkasad) [reply ui.go:417]: Yes, it changes the appearance when the window width is less than the text width. Instead of the text getting cut off, it now wraps.

When `w.Label` was used, `"CB"` defined the text alignment, with the `C` meaning centered and the `B` meaning the bottom of the label.

`w.LabelWrap` doesn't support custom alignment. Text will be aligned to the top-left, because it makes implementing wrapping much easier. So it would be equivalent to `"LT"` instead of `"CB"` when the text fits within the window.

Also see my latest comment for screenshots.

**Comment by kdkasad** [comment]
# Screenshots

With the old behavior, text is centered when it fits into the window entirely. When it didn't fit, the text was cut off and part of the message is completely unreadable.

With the new behavior, text is left-aligned always.* When it doesn't fit in the window, the text is wrapped onto new lines so that it remains readable.

*The title is still centered. Long titles will still be cut off if they don't fit in the window.

## Old
### Capabilities view
Large window:
large window

Small window (text cuts off):
small window

### Error view
Large window:
large window

Small window:
small window

## New
### Capabilities view
Large window (text is not centered):
large window

Small window (text wraps):
small window

### Error view
Large window:
large window

Small window:
small window

File: ui.go
Comment (kdkasad) [reply ui.go:498]: Done in eded3a2

File: cli.go
Comment (AXDOOMER) [reply cli.go:32]: You don't need to change it back. I don't believe it breaks anything. If most programs use `-v`, then we'll use it.

Comment (kdkasad) [reply cli.go]: Done in 2d49fe1

File: cli.go
Comment (AXDOOMER) [reply cli.go:188]: No, that's fine. We don't _need_ to make the code different, we just need a clean room implementation without looking at the original code. If two lines are the same in an area where there aren't many implementations possible, then that's just a coincidence.

File: module.go
Comment (AXDOOMER) [reply module.go:192]: Yes, we'll leave it this way. I don't know which other errors could happen here, NoiseTorch has many error checks for all sorts of things (which is a good thing), but I don't have a system with the flaws required to trigger those error checks and see if those error conditions work correctly.

File: cli.go
Comment (TheDukeofErl) [reply cli.go]: Perhaps it would be best to have it as `-v` or `-l`? Generally, when I think about cli opts, something with a "-" should have a single character after it, while something with a longer name has a "--". I suppose that's a long way of saying that I'm okay with this changing to `-v' or even '-l'.

**Comment by TheDukeofErl** [comment]
> # Screenshots
> ...
> Edit: now I'm wondering if the "title" on the capabilities view should be a wrapped label, since it's fairly long. This does have the downside of making it left-aligned, though.

I'm fine with the text being left aligned and wrapping. With that in mind, I do think it would be best to have consistency here: I would prefer the title, for example "Error", to also be left aligned.

File: cli.go
Comment (kdkasad) [reply cli.go]: > You don't need to change it back. I don't believe it breaks anything. If most programs use `-v`, then we'll use it.

>  I suppose that's a long way of saying that I'm okay with this changing to `-v' or even '-l'.

Ok sounds good, I'll remove the 2d49fe1 and leave it as `-v`.

> Generally, when I think about cli opts, something with a "-" should have a single character after it, while something with a longer name has a "--".

I completely agree.

**Comment by AXDOOMER** [comment]
> **Edit:** now I'm wondering if the "title" on the capabilities view should be a wrapped label, since it's fairly long. This does have the downside of making it left-aligned, though. (That's if wrapping ends up being approved)

I suggest you edit the window settings of this program and enforce a minimum window size. I don't know if that's doable. This would avoid many UI glitches that happen when the window is too small. A minimum resolution could probably be 800x640? I think it would fit most modern displays. 


> I'm fine with the text being left aligned and wrapping. With that in mind, I do think it would be best to have consistency here: I would prefer the title, for example "Error", to also be left aligned.

I disagree. Because it's the title, I don't think consistency applies here. I also believe that most things should be centered because it looks better (the exit status for example) as long as its not so long that it would require a line warp at the minimum window size that we would set.

**Comment by kdkasad** [comment]
> > **Edit:** now I'm wondering if the "title" on the capabilities view should be a wrapped label, since it's fairly long. This does have the downside of making it left-aligned, though. (That's if wrapping ends up being approved)
> 
> I suggest you edit the window settings of this program and enforce a minimum window size. I don't know if that's doable. This would avoid many UI glitches that happen when the window is too small. A minimum resolution could probably be 800x640? I think it would fit most modern displays.

This is an option I hadn't thought of. The reason I noticed that text is cut off is because I was using a tiling WM where the window size is managed by the WM, not the application. So for those using tiling WMs, this solution wouldn't work as the minimum window size is often ignored.

> > I'm fine with the text being left aligned and wrapping. With that in mind, I do think it would be best to have consistency here: I would prefer the title, for example "Error", to also be left aligned.
> 
> I disagree. Because it's the title, I don't think consistency applies here.

I have to agree with @AXDOOMER on this one. I don't think the title on error views should be centered.

I think the best solution would be one where if the window is above certain width, the text not wrap and will be centered (original behavior). Then if the window is smaller than this certain width, the text will wrap and be left-aligned (new behavior).

However, I propose we defer this UI debate to a new issue and simply decide whether to keep or drop 2a7ec81 in this PR.

**Comment by TheDukeofErl** [comment]
> Because it's the title, I don't think consistency applies here.

Could we just shorten the title here then? I think the current one is more wordy than it needs to be: I think that something along the lines of "Missing capabilities" as a title would tell enough of the story. Potentially, we could consider adding additional explanation to the message itself, rather than the title. For example, "This program requires "CAP_SYS_RESOURCE" in order to function properly. ... <same as rest of message>."

File: ui.go
Comment (AXDOOMER) [reply ui.go:498]: Please add `v` before `%d.%d.%d` so it's consistent with `v0.3.28`.

**Review by AXDOOMER — CHANGES_REQUESTED** [review]

**Comment by kdkasad** [comment]
> > Because it's the title, I don't think consistency applies here.
> 
> Could we just shorten the title here then? I think the current one is more wordy than it needs to be: I think that something along the lines of "Missing capabilities" as a title would tell enough of the story. Potentially, we could consider adding additional explanation to the message itself, rather than the title. For example, "This program requires "CAP_SYS_RESOURCE" in order to function properly. ... ."

That's definitely viable. I was thinking about just adding a short title above the current text and pushing the rest down. That way, we can wrap the long lines and have a centered title without much worry about it cutting off.

**Comment by TheDukeofErl** [comment]
> However, I propose we defer this UI debate to a new issue and simply decide whether to keep or drop [2a7ec81](https://github.com/noisetorch/NoiseTorch/commit/2a7ec816538f125ed63107b578a5ec582b540770) in this PR.

I'm fine with keeping the change as is and resolving it in a new issue as well. There's a lot of ways to skin the cat in this case.

**Comment by AXDOOMER** [comment]
I think for now you could just make "Missing capabilities" the title of this window and make other longer lines warp.

**Comment by kdkasad** [comment]
> I think for now you could just make "Missing capabilities" the title of this window and make other longer lines warp.

Ok, settling on this for now. Any other changes relating to text wrapping will be deferred to a new issue.

**Comment by kdkasad** [comment]
## New screenshots
Large window:
Screenshot from 2022-09-02 21-50-38

Small window:
Screenshot from 2022-09-02 21-50-42

For some weird reason, the row height must be >=50 in order for the text to actually wrap, even though it doesn't use all of this space and would probably fit into a height of 30.

**Comment by kdkasad** [comment]
Still don't know why re-requesting a review from @AXDOOMER removes the review request for @TheDukeofErl. Re-reviews from both are welcome.

**Review by AXDOOMER — APPROVED** [review]
I think we're complete at this point. We'll wait for @TheDukeofErl 's approval before merging.

**Comment by kdkasad** [comment]
> ...if you'd like to continue, I'll proceed with removing some of the Pipewire code from NoiseTorch's source code that must be rewritten. It's important that you do not look at the Pipewire code and that you have not already looked at it so that it can be a clean room reimplementation.

I have looked at the PipeWire code a little bit, though mostly just the logic for getting node (sink/source) names so that I could implement #358.

Unfortunately, I think that probably disqualifies me for a while.

**Comment by navarro967** [comment]
@kdkasad, Amazing work so far!

@AXDOOMER @kdkasad, I would be willing to take a crack at the Pipewire re-implementation. Will there be a new branch for this?

**Comment by AXDOOMER** [comment]
> I would be willing to take a crack at the Pipewire re-implementation. Will there be a new branch for this?

I will @ you when ready with the instructions.

**Review by TheDukeofErl — APPROVED** [review]
LGTM

**Comment by ZyanKLee** [comment]
awesome work, folks!

**Pull request merged**
